staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -21,56 +21,128 @@ import (
-	// slices described in the driver resources parameters are valid
-	// according to the restrictions defined in the resource.k8s.io API.
+	// At the time when this gets called, the original ResourceClaims may have
+	// been deleted already. They also don't get cached by the kubelet. Therefore
+	// parameters for each ResourcClaim are only the UID, namespace and name.
+	// It is the responsibility of the DRA driver to cache whatever additional
+	// information it might need about prepared resources.
 	//
-	// Invalid ResourceSlices will be rejected by the apiserver during
-	// publishing, which happens asynchronously and thus does not
-	// get returned as error here. The only error returned here is
-	// when publishing was not set up properly, for example missing
-	// [KubeClient] or [NodeName] options.
+	// This call must be idempotent because the kubelet might have to ask
+	// for un-preparation multiple times, for example if it gets restarted.
+	// Therefore it is not an error if this gets called for a ResourceClaim
+	// which is not currently prepared.
 	//
-	// The caller may modify the resources after this call returns.
-	PublishResources(ctx context.Context, resources resourceslice.DriverResources) error
+	// As with PrepareResourceClaims, the helper takes care of logging
+	// and serializaton.
+	//
+	// The conventions for returning one overall error and several per-ResourceClaim
+	// errors are the same as in PrepareResourceClaims.
+	UnprepareResourceClaims(ctx context.Context, claims []NamespacedObject) (result map[types.UID]string, err error)
+}
+
+// PrepareResult contains the result of preparing one particular ResourceClaim.
+type PrepareResult struct {
+	// Error, if non-empty, describes a problem that occured while preparing
+	// the ResourceClaim. The devices are then ignored and the kubelet will
+	// try to prepare the ResourceClaim again later.
+	Error string
 
-	// This unexported method ensures that we can modify the interface
-	// without causing an API break of the package
-	// (https://pkg.go.dev/golang.org/x/exp/apidiff#section-readme).
-	internal()
+	// Devices contains the IDs of CDI device associated with specific requests
+	// in a ResourceClaim. Those IDs will be passed on to the container runtime
+	// by the kubelet.
+	//
+	// The empty slice is also valid.
+	Devices []Device
+}
+
+// Device provides the CDI device IDs for one request in a ResourceClaim.
+type Device struct {
+	// Requests lists the names of requests or subrequests in the
+	// ResourceClaim that this device is associated with. The subrequest
+	// name may be included here, but it is also okay to just return
+	// the request name.
+	//
+	// A DRA driver can get this string from the Request field in
+	// [resourceapi.DeviceRequestAllocationResult], which includes the
+	// subrequest name if there is one.
+	//
+	// If empty, the device is associated with all requests.
+	Requests []string
```

Comment:
@mortent: this is what the helper expects from the DRA driver. Note that it doesn't ask DRA driver authors to strip the subrequest name. Instead, the helper will do that before returning this as result to the kubelet.

Looks okay to you?

Reply:
Yeah, that makes sense to me. Since we don't allow references to subrequests in the pod/container specs, the logic that determines which requests should be available to pods/containers must strip away the subrequest (if specified). Doing this in the helper seems like the right approach.